python/metatensor_torch/metatensor/torch/atomistic/ase_calculator.py

```diff
@@ -501,31 +527,96 @@ def compute_energy(
             length_unit="angstrom",
             outputs=outputs,
         )
-        energies = self._model(
+        predictions = self._model(
             systems=systems,
             options=options,
             check_consistency=self.parameters["check_consistency"],
-        )["energy"]
+        )
+        energies = predictions["energy"]
 
         results_as_numpy_arrays = {
             "energy": energies.block().values.detach().cpu().numpy().flatten().tolist()
         }
         if compute_forces_and_stresses:
-            energy_tensor = energies.block().values
-            energy_tensor.backward(torch.ones_like(energy_tensor))
-            results_as_numpy_arrays["forces"] = [
-                -system.positions.grad.cpu().numpy() for system in systems
-            ]
-            if all(atoms.pbc.all() for atoms in atoms_list):
-                results_as_numpy_arrays["stress"] = [
-                    strain.grad.cpu().numpy() / atoms.cell.volume
-                    for strain, atoms in zip(strains, atoms_list)
+            if self._non_conservative:
+                results_as_numpy_arrays["forces"] = (
+                    predictions["non_conservative_forces"]
+                    .block()
+                    .values.squeeze(-1)
+                    .detach()
+                    .cpu()
+                    .numpy()
+                )
+                # all the forces are concatenated in a single array, so we need to
+                # split them into the original systems
+                split_sizes = [len(system) for system in systems]
+                split_indices = np.cumsum(split_sizes[:-1])
+                results_as_numpy_arrays["forces"] = np.split(
+                    results_as_numpy_arrays["forces"], split_indices, axis=0
+                )
```

Comment:
We have the metadata to do the split in the samples … This should be a call to `mts.split()` (and a good use case for #627)

Reply:
It's slower though and it doesn't simplify the code in this case

Reply:
Fair enough for the speed, although I'd be interested in updating `split` to take advantage of #831